csrc/scheduler/reduction.cpp

```diff
@@ -521,15 +521,32 @@ std::shared_ptr<ReductionParams> outerReductionHeuristic(
   const int64_t device_multiprocessor_count =
       (int64_t)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
 
+  // Set register used to store vectorized and unrolled data loaded from gmem.
+  // A large value allows more unroll and vectorization, which is beneficial
+  // for memory-bound kernels. However, it increases register pressure and may
+  // lead to lower occupancy which is bad for compute-bound kernels. In most
+  // cases, the scheduler uses 512 threads and to reach an occupancy of 50%,
+  // each thread can use up to 64 registers, here only 8 registers are reserved
+  // for unroll and vectorization. The fused ops can have 48 registers for other
+  // purposes. Test shows it leads to 50% occupancy for outer reduction without
+  // fused ops and 50% occupancy for gelu backward which fused 21 ops including
+  // the expensive tanh op. Further tuning of this heuristic can utilize the
+  // cost of the fused ops.
+  const int64_t buffer_reg_count = 8L;
   auto const max_unroll = ceilDiv(
       // Available unrolling based on size of data type
-      (int64_t)16 / (int64_t)max_input_dtype_size,
+      buffer_reg_count * scheduler_utils::bytes_per_register /
+          (int64_t)max_input_dtype_size,
       // Reduce unrolling if we have many inputs, start reduction at 4 inputs
       scheduler_utils::lastPow2(
           std::max((int64_t)n_tensor_inputs >> 2, (int64_t)1)));
 
   const int64_t n_elems = total_reduction_numel * total_iteration_numel;
-  const int64_t n_waves = 8;
+  // Try to use 4 * SM blocks to reduce communication cost. But still
+  // use 8 * SM blocks if the problem size is large, it increased requested
+  // waves, helps hiding memory latency, if still use 4 * SM blocks, about 5%
+  // regression for compute bound kernels. Not much change for memory bound.
+  const int64_t n_waves = n_elems >= (int64_t)64 * 1024 * 1024 ? 8 : 4;
```

Comment:
nitpick, this is also contradicting PR description, which were saying that we are always targeting 4 waves.

Let's have a loud comment on this number being justified empirically.

Reply:
revised PR description.